src/viper/README.md

```diff
@@ -161,23 +163,121 @@ Below, we summarize the language features that _Motoko-san_ currently supports.
+              case null ...;
+              case (?a) ...;
+          }
+          ```
+
+          Nested matching is _not_ supported, i.e. the subpattern `a` in `?a`
+          must be a variable pattern.
+
+    * Variant types `{ #A; #B : T₁; #C : (T₂, T₃) }` (with limitations)
+        * Construction: `#A`, `#B(x)`, `#C(x, y)`
+        * Deconstruction by pattern matching:
 
-  * `Int`: `+`, `/`, `*`, `-`, `%`
+          ```motoko
+          switch x {
+              case #A ...;
+              case #B(x) ...;
+              case #C(x, y) ...;
+          }
+          ```
 
-  * Relations: `==`, `!=`, `<`, `>`, `>=`, `<=`
+          Nested matching is _not_ supported, i.e. the subpatterns `x`, `y` in
+          `#B(x)` and `#C(x, y)` must all be variable patterns.
 
-  Supporting `Text`, `Nat`, `Int32`, tuples, record, and variants is _simple_.
+        * Generic and recursive variants are supported, e.g.
 
-    Supporting `Option<T>` types is _trivial_.
+          ```motoko
+          type List<T> = { #Nil; #Cons : (T, List<T>) }
+          ```
+
+        * Limitation 1: Nominative equality. In the Viper translation, equality
+          of variant types is not _structural_ (as in normal Motoko programs)
+          but _nominative_.
+          * One consequence of this is that a variant type may not be used
+            directly but must be bound to a name:
+
+            ```motoko
+            type N = { #A; #B }
+            ```
+
+            Instead of using `{ #A; #B }` in type signatures, one must always
+            refer to this variant by its name `N`.
+          * Furthermore, given two identical variant declarations, they are
+            treated as distinct in the translation:
+
+             ```motoko
+             type N = { #A; #B }
+             type M = { #A; #B }   // M ≠ N
+             ```
+
+          * Motoko's type checker has not been modified to account for
+            nominative equality of variants, so type inference is not reliable:
+            if a variant is bound to a variable in `let`, `var`, or `case`, the
+            bound variable must be explicitly annotated with the named type of
+            the variant.
+        * Limitation 2: Tuple components. If the type associated with a
+          constructor is a tuple, its components must be specified individually.
+          That is:
+
+          ```motoko
+          type N = { #Con: (Int, Bool) }
+
+          let x = #Con(10, true)   // supported
+
+          let p = (10, true)
+          let x = #Con(p)          // not supported
+          ```
+
+    Supporting `Text`, `Int32`, tuples, record, and variants is _simple_.
 
     Supporting `async` types is _hard_.
 
     Supporting `Float`, function types, co-inductive, mutually recursive, and sub-typing is _hard_.
 
-    Supporting container types and generic types, e.g., arrays (`[var T]`) and `HashMap<K, V>`, is _hard_.
+    Supporting container types and generic types, e.g. `HashMap<K, V>`, is _hard_.
```

Comment:
Do you think this is conceptually harder than what we do for Arrays, or about the same complexity?

Reply:
If we find the right axioms, it should be possible, but we won't know until we try. For arrays there was an example in Viper's documentation, a hashmap would require more tinkering

Reply:
Thanks, this is just my curiosity.

We do have some drafts for translating HashMap from Motoko to Viper from a very long time ago (even before Motoko-san was created). But let's keep working on concrete examples, and I don't think we need to add anything with HashMaps into this project (there's enough interesting canisters with arrays, ADTs and things that we already support — we should utilize those before expanding).